Utilities for ease data transmission trough connection

AgainPsychoX opened the pull request:
Here is draft pull request that address issue #41 (goals described there, check them out).

To do:
- [x] basic packets - 2 bytes of type, 2 bytes of length and max of 65536 bytes of data,
- [x] basic testing (multiple packets sent trough, also trying to fire tons of these),
- [x] update chat app example using the packets system _(also written nicely in Javascript)_,
- [x] add more features to the chat app example to show why such packet system is good,
    - [x] user joined/left,
    - [x] remove message,
    - [x] seen by all message,
    - [x] mute/unmute,
    - [x] some permission (like removing only user own messages),
    - [x] edit message,
    - [x] flood warning,
- [ ] ensured packets (with checksum; i.e. it would Arduino with HC-05 work even while there are some noises between these), _This might be included in some other PR to don't block more important things._
- [ ] wide testing (multiple client chat room 🤔 )

_For now it is Work-In-Progress pull request to get feed back._

AgainPsychoX committed:
Basic packet communication with simplest example

AgainPsychoX committed:
Use new packets system in Chat example

+ Example server script now allow multiple connections,
+ User joined/left messages.
- Now we lack some simplest raw serial communication example :/

AgainPsychoX committed:
Minor fixes and improvements for Chat example

+ Use 2 bytes for message ID now.
+ Fix not true throw `not implemented` while removing message.

AgainPsychoX committed:
Add different colors for multiple users

/shrug command introduced :D

AgainPsychoX committed:
Add `message seen` to chat example

+ Fix in some utils in example chat server script.
* Noticed not reliable behavior while fast spamming multiple packets (need further test to check is it a case for previous versions).

AgainPsychoX committed:
Add `mute`, `unmute` and `kick` commands to Chat example server.

PsychoXIVI commented:
Adding multiple features overtime, but to have in mind: Most of these need more testing (especially multiple clients situations).

AgainPsychoX committed:
Fix packets corrupting while spamming

+ Add `addIterable` to `BluetoothConnection``.output` which guarantees to send data sequentially.

AgainPsychoX committed:
Fix `message seen` of chat example

+ Minor fix in broadcasting packets (use public `sendPacket`of `Client`) in example chat server script.

AgainPsychoX committed:
Implement flooding warnings on example chat server

AgainPsychoX committed:
Remove messages on client side in case of flooding (chat example)

AgainPsychoX committed:
Minor changes in example chat server script

+ Move muted and flooding checks above reading message (in `PushMessage` packet handler).
+ Add unknown server console command warning.

AgainPsychoX committed:
Minor fixes in chat example

+ Fix missing high byte of message packet (index by-one-miss issue).
+ Do proper disposing of text editing and list scroll controllers in app.

AgainPsychoX committed:
Implement message editing in chat example

Note: still missing few checks.

AgainPsychoX committed:
Add simple messages history to track message ownership for chat example

Used in checks for removing and editing messages.

AgainPsychoX committed:
Update flooding and muted checks order in chat example server script

Now spamming while being muted also ends up in being kicked from the server.

PsychoXIVI commented:
Some wider tests required, and then it might be ready to merge. 

Also, I noticed some problems related to interactions between iterators, decoders and etc. I will look into these (also in project wide code base), but it will not prevent this pull request.

AgainPsychoX committed:
Fixes for chat example app (after multiple client tests)

There are some fixes after testing with two clients (first time for me :/ I borrowed another Android phone to do so).

+ Proper new client information collection.

+ Fix a bit "seen by all" feature.
Now it works properly in multi-user session, but does not in single user only - which is acceptable, since it's chat app example, not a diary ;)

+ Update size of text containers using `MediaQuery` to fix on all devices.
+ Remove some old debug print.

AgainPsychoX committed:
Add `UserInfo`packet to chat app example

...to make fresh clients know about these already connected, instead leaving those as "ghosts" with gray color. 

Keep in mind this is simple chat app/server example. In real world such packet should be aggregated with many users informations at once, instead of one separated packet per one user.

PsychoXIVI commented:
I came to conclusion, that Dart lacks some elements that could make handling packets (and parsing strings and etc.) more efficient, without unnecessary copying and so on.

As for now, I think there is no way to do something about it, since there is much more things in-deep... Just imagine:
* Bluetooth driver writes to kernel provided buffer (after socket was opened), then 
* the protocol implementation strips the frames/packets from its headers, copying ready-to-read data to again other buffer,
* the application reads from the buffer to user-land buffer, then 
* the plugin implementation packs readings into `Uint8List` objects, then
* the user of plugin converts those objects like decoding it to strings.

In ideal world, there would be need only for two copying at most instead of like four. Drivers should interpret incoming data against protocols and then write to multiple prepared buffers (each separate for port/socket). The buffer should be readable from user-app-land, might be realized by special inteligent iterator or some object - to keep track of status and position in buffer (which is not endless, so in some point oldest data would be discarded). Here, it is important to mention how big is virtual memory addressing - it would be easy to Then, Flutter would allow pass and use this iterator inside user application, then the plugin user would use this iterator, like feeding it into decoder in order to parse string (it would be second copying).
